File: pe/nonvisualstudio/elixir/Implementation/Problem0028.ex
defmodule Problem0028 do

  def sumOfBoxLevel(n), do: _sumOfBoxLevel(0, 0, n)

  def sumOfSpiralDiagonals(sizeOfSpiral), do: _sumOfSpiralDiagonals(sizeOfSpiral, 1)

  defp _sumOfSpiralDiagonals(1, sumSoFar), do: sumSoFar
  defp _sumOfSpiralDiagonals(currentLevel, sumSoFar) do
    nextCurrentLevel = currentLevel - 2
    nextSumSoFar = sumSoFar + sumOfBoxLevel(currentLevel)

    _sumOfSpiralDiagonals(nextCurrentLevel, nextSumSoFar)
  end

  defp _sumOfBoxLevel(result, 4, _), do: result
  defp _sumOfBoxLevel(result, i, n) do
    nextI = i + 1
    nextResult = (n * n) - ((n - 1) * i) + result

    _sumOfBoxLevel(nextResult, nextI, n)
  end
end

File: pe/nonvisualstudio/clojure/Implementation/Problem0028.clj
(ns pe.clojure.implementation.problem0028)


(defn sumOfBoxLevel [n]
  (loop [i 0
         result 0]
    (cond
     (= i 4) result
     :else (recur
            (inc i)
            (+ result
             (- (* n n)
                (* (- n 1) i)))))))


(defn sumOfSpiralDiagonals [sizeOfSpiral]
  (loop [i sizeOfSpiral
         sumSoFar 1]
    (if (= i 1)
      sumSoFar
      (recur
       (- i 2)
       (+ sumSoFar
          (sumOfBoxLevel i))))))
